actors/ppo_atari_pz.py

```diff
@@ -0,0 +1,1085 @@
+        "image_size": [6, 84, 84],
+        "buffer_capacity": 100_000,
+    }
+
+    def __init__(self, environment_specs: EnvironmentSpecs, cfg: EnvironmentConfig) -> None:
+        super().__init__()
+        self._dtype = torch.float32
+        self._environment_specs = environment_specs
+        self._cfg = cfg
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.returns = 0
+        self.model = PPOModel(
+            model_id="",
+            environment_implementation=self._environment_specs.implementation,
+            num_actions=utils.flatdim(self._environment_specs.get_action_space().gym_space),
+            input_shape=tuple(self._cfg.image_size),
+            num_policy_outputs=1,
+            learning_rate=self._cfg.learning_rate,
+            n_iter=self._cfg.num_epochs,
+            device=self._cfg.device,
+            dtype=self._dtype,
+        )
+
+    @abstractmethod
+    async def trial_sample_sequences_producer_impl(self, sample_producer_session: SampleProducerSession):
+        raise NotImplementedError
+
+    @abstractmethod
+    async def impl(self, run_session: RunSession) -> dict:
+        raise NotImplementedError
+
+    async def train_step(
+        self,
+        observations: List[torch.Tensor],
+        rewards: List[torch.Tensor],
+        actions: List[torch.Tensor],
+        dones: List[torch.Tensor],
+    ):
+        """Train the model after collecting the data from the trial"""
+
+        # Take n steps from the rollout
+        observations = torch.vstack(observations).to(self._device)
+        actions = torch.vstack(actions).to(self._device)
+        rewards = torch.vstack(rewards).to(self._device)
+        dones = torch.vstack(dones).to(self._device)
+
+        # Make a dataloader in order to process data in batch
+        batch_state = self.make_dataloader(observations[:-1], self._cfg.batch_size, self.model.input_shape)
+        batch_action = self.make_dataloader(actions, self._cfg.batch_size, (self.model.num_policy_outputs,))
+
+        values = self.compute_value(batch_state)
+        with torch.no_grad():
+            next_value = self.model.network.get_value(observations[-1:])
+        next_value = next_value * (1 - dones[-1])
+        values = torch.cat((values, next_value), dim=0)
+        log_probs = self.compute_batch_log_lik(batch_state, batch_action)
+
+        # Compute the generalized advantage estimation
+        advs = self.compute_gae(
+            rewards=rewards, values=values, dones=dones, gamma=self._cfg.discount_factor, lam=self._cfg.lambda_gae
+        )
+
+        # Update parameters for policy and value networks
+        policy_loss, value_loss = self.update_parameters(
+            observations=observations[:-1],
+            actions=actions,
+            advs=advs,
+            values=values[:-1],
+            log_probs=log_probs,
+            num_epochs=self._cfg.num_epochs,
+        )
+
+        return policy_loss, value_loss
```

Comment:
I would suggest an alternatve version of the `train_step`. The reason is that the code will not work if you run more than 1 trial before doing a `train_step`. The reason for that is that you always need an extra sample at the end of the trial to compute the advantage values, but you need to remove it when updating the parameters (ex: `observations[:-1]`, `values[:-1]`). This is a mask that removes the last sample. 

Instead I suggest a mask that removes the last sample of every trial done, no matter how many you run before a training step. The `obs_mask` will track all samples with `done` flag, still use them to compute the advantages, but remove them when updating parameters.

```suggestion
    async def train_step(
        self,
        observations: List[torch.Tensor],
        rewards: List[torch.Tensor],
        actions: List[torch.Tensor],
        dones: List[torch.Tensor],
    ):
        """Train the model after collecting the data from the trial"""

        # Take n steps from the rollout
        _observations = torch.vstack(observations).to(self._device)
        _actions = torch.vstack(actions).to(self._device)
        _rewards = torch.vstack(rewards).to(self._device)
        _dones = torch.vstack(dones).to(self._device)

        # Normalize the observations
        if self.model.state_normalization is not None:
            self.model.state_normalization.update(_observations)

        # Mask to remove values and observations from done samples.
        mask = _dones == 0
        obs_mask = mask.repeat(1, _observations.shape[1])
        masked_observations = _observations[obs_mask].reshape([_actions.shape[0], _observations.shape[1]])

        # Make a dataloader in order to process data in batch
        adv_batch_state = self.make_dataloader(_observations, self._cfg.batch_size, self.model._num_input)

        values = self.compute_value(adv_batch_state)
        next_values = values * (1 - _dones)

        # Compute the generalized advantage estimation
        advs = self.compute_gae(
            rewards=_rewards,
            values=next_values,
            dones=_dones,
            gamma=self._cfg.discount_factor,
            lam=self._cfg.lambda_gae,
        )

        batch_state = self.make_dataloader(masked_observations, self._cfg.batch_size, self.model._num_input)
        batch_action = self.make_dataloader(_actions, self._cfg.batch_size, self.model._num_output)
        log_probs = self.compute_batch_log_lik(batch_state, batch_action)

        # Update parameters for policy and value networks
        policy_loss, value_loss = self.update_parameters(
            observations=masked_observations,
            actions=_actions,
            advs=advs,
            values=values[mask],
            log_probs=log_probs,
            num_epochs=self._cfg.num_epochs,
        )

        return policy_loss, value_loss

    def update_parameters(
        self,
        observations: torch.Tensor,
        actions: torch.Tensor,
        advs: torch.Tensor,
        values: torch.Tensor,
        log_probs: torch.Tensor,
        num_epochs: int,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        """Update policy & value networks"""

        returns = advs + values
        num_obs = len(returns)
        global_idx = np.arange(num_obs)
        for _ in range(num_epochs):
            np.random.shuffle(global_idx)
            for i in range(0, num_obs, self._cfg.batch_size):
                idx = global_idx[i : i + self._cfg.batch_size]
                if len(idx) < self._cfg.batch_size:
                    break
                state = observations[idx].to(self._device)
                action = actions[idx].to(self._device)
                return_ = returns[idx].to(self._device)
                adv = advs[idx].to(self._device)
                old_log_prob = log_probs[idx].to(self._device)

                # Compute the value and values loss
                value = self.model.value_network(state)
                value_loss = (
                    torch.nn.functional.mse_loss(return_, value.repeat(1, return_.shape[1])) * self._cfg.value_loss_coef
                )

                # Get action distribution & the log-likelihood
                action_dist, _ = self.model.policy_network(state)
                new_log_prob = action_dist.log_prob(action)
                ratio = torch.exp(new_log_prob - old_log_prob)

                # Compute policy loss
                policy_loss_1 = adv * ratio
                policy_loss_2 = adv * torch.clamp(ratio, 1 - self._cfg.clipping_coef, 1 + self._cfg.clipping_coef)
                policy_loss = -torch.min(policy_loss_1, policy_loss_2).mean()

                # Update value network
                self.model.value_optimizer.zero_grad()
                value_loss.backward()
                torch.nn.utils.clip_grad_norm_(self.model.value_network.parameters(), self._cfg.grad_norm)
                self.model.value_optimizer.step()

                # Udapte policy network
                self.model.policy_optimizer.zero_grad()
                policy_loss.backward()
                torch.nn.utils.clip_grad_norm_(self.model.policy_network.parameters(), self._cfg.grad_norm)
                self.model.policy_optimizer.step()

        # Decaying learning rate after each update
        self.model.policy_scheduler.step()
        self.model.value_scheduler.step()

        return policy_loss, value_loss
```

Reply:
@wduguay-air Awesome. I will test it out and push the new commit

Reply:
@wduguay-air your code breaks when running the training. The reason is that the observation shape in this case is `(B, C, H, W)` where C is image channel, H is height of the image, and W is the width of the image. These lines of code are no longer working
```Python
obs_mask = mask.repeat(1, _observations.shape[1])
masked_observations = _observations[obs_mask].reshape([_actions.shape[0], _observations.shape[1]])
```
I will give a try to fix it...